File: InMotionGIT.Common/Extensions/DataRowExtensions.vb
#Region "using"

Imports System.IO
Imports System.Runtime.CompilerServices

#End Region

Namespace Extensions

    ''' <summary>
    ''' Extension methods for the DataRow type
    ''' </summary>
    Public Module DataRowExtensions

#Region "Binary (BLOB) Extension"

        <Extension()>
        Public Function FileContent(value As DataRow, name As String, path As String, filename As String, extension As String) As String
            Dim result As String = String.Empty

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                Dim localfilename As String = path
                If filename.IsEmpty Then
                    localfilename &= "\" & Guid.NewGuid().ToString()
                Else
                    localfilename &= "\" & filename
                End If
                If extension.StartsWith(".") Then
                    localfilename &= extension
                Else
                    localfilename &= "." & extension
                End If
                localfilename = localfilename.Replace("\\", "\")

                Using FS As New FileStream(localfilename, FileMode.Create)
                    Dim blob As Byte() = DirectCast(value(name), Byte())
                    FS.Write(blob, 0, blob.Length)
                    FS.Close()
                End Using

                result = localfilename

            End If

            Return result
        End Function

#End Region

#Region "Numeric Extension"

        <Extension()>
        Public Function NumericValue(value As DataRow, name As String) As Decimal
            Dim result As Decimal = 0

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
            End If

            Return result
        End Function

        <Extension()>
        Public Function IntegerValue(value As DataRow, name As String) As Integer
            Dim result As Integer = 0

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
            End If

            Return result
        End Function

        <Extension()>
        Public Function ByteValue(value As DataRow, name As String) As Byte
            Dim result As Byte = 0

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
            End If

            Return result
        End Function

        <Extension()>
        Public Function NumericValueWithFormatDefault(value As DataRow, name As String, format As String, defaultValue As Object, ByRef specified As Boolean) As Object
            Dim result As Object = defaultValue

            If value.IsNotNull(name) Then
                Dim internal As Decimal = NumericValue(value, name)

                If format.IsNotEmpty Then
                    result = internal.ToString(format, New System.Globalization.CultureInfo("en-US", False))
                Else
                    result = internal
                End If

                specified = True
            Else
                specified = False
            End If

            Return result
        End Function

#End Region

#Region "Double Extension"

        <Extension()>
        Public Function DoubleValue(value As DataRow, name As String) As Double
            Dim result As Double = 0

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
            End If

            Return result
        End Function

#End Region

#Region "DateTime Extension"

        <Extension()>
        Public Function DateTimeValue(value As DataRow, name As String) As Date
            Dim result As Date = Date.MinValue

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
            End If

            Return result
        End Function

        <Extension()>
        Public Function DateTimeValueWithFormatDefault(value As DataRow, name As String, format As String, defaultValue As Object) As Object
            Dim result As Object = defaultValue

            If value.IsNotNull(name) Then
                Dim internal As Date = DateTimeValue(value, name)

                If format.IsNotEmpty Then
                    result = internal.ToString(format)
                Else
                    result = internal
                End If
            End If

            Return result
        End Function

#End Region

#Region "Boolean Extension"

        <Extension()>
        Public Function BooleanValue(value As DataRow, name As String) As Boolean
            Dim result As Boolean = 0

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = (value.Item(name) = 1)
            End If

            Return result
        End Function

        <Extension()>
        Public Function BooleanCharValue(value As DataRow, name As String) As Boolean
            Dim result As Boolean = 0

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = (value.Item(name) = "1")
            End If

            Return result
        End Function

#End Region

#Region "String Extension"

        <Extension()>
        Public Function StringValue(value As DataRow, name As String) As String
            Dim result As String = String.Empty

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
                result = result.Trim
            End If

            Return result
        End Function

        <Extension()>
        Public Function StringValueWithDefault(value As DataRow, name As String, defaultValue As Object, ByRef specified As Boolean) As String
            Dim result As String = defaultValue

            If value.IsNotNull(name) Then
                result = StringValue(value, name)
                specified = True
            Else
                specified = False
            End If

            Return result
        End Function

        <Extension()>
        Public Function StringHourValue(value As DataRow, name As String) As Date
            Dim result As Date = Date.MinValue
            Dim currentHour As Integer = 0
            Dim currentMinute As Integer = 0
            Dim currentValue As String

            If Not IsNothing(value.Item(name)) AndAlso
           Not IsDBNull(value.Item(name)) Then
                currentValue = value.Item(name)
                currentValue = currentValue.Trim
                If currentValue.IndexOf(":") > 0 Then
                    currentHour = currentValue.Split(":")(0)
                    currentMinute = currentValue.Split(":")(1)
                End If
            End If

            result = result.AddHours(currentHour)
            result = result.AddMinutes(currentMinute)

            Return result
        End Function

#End Region

#Region "Specials Extension"

        <Extension()>
        Public Function GuidValue(value As DataRow, name As String) As Guid
            Dim result As Guid = Guid.NewGuid

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = value.Item(name)
            End If

            Return result
        End Function

        <Extension()>
        Public Function XmlValue(value As DataRow, name As String) As XDocument
            Dim result As New XDocument

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = XDocument.Parse(value.StringValue(name))
            End If

            Return result
        End Function

        <Extension()>
        Public Function SwitchValue(value As DataRow, name As String) As Boolean
            Dim result As Boolean = False

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = (value.Item(name) = 1)
            End If

            Return result
        End Function

        <Extension()>
        Public Function SwitchCharValue(value As DataRow, name As String) As Boolean
            Dim result As Boolean = False

            If Not IsNothing(value.Item(name)) AndAlso Not IsDBNull(value.Item(name)) Then
                result = (value.Item(name) = "1")
            End If

            Return result
        End Function

#End Region

#Region "Behavior Functions"

        <Extension()>
        Public Function IsNotNull(value As DataRow, name As String) As Boolean
            Return Not IsDBNull(value.Item(name))
        End Function

        <Extension()>
        Public Function EnumValue(Of T)(ByVal value As DataRow, name As String, ByVal enumType As T, ByRef specified As Boolean, ByRef witherror As Boolean) As T

            Dim result As T = Nothing

            If value.IsNotNull(name) Then
                Try
                    result = CType([Enum].Parse(enumType.GetType, value.Item(name)), T)
                    specified = True
                Catch ex As ArgumentException
                    result = Nothing
                    specified = False
                    witherror = True
                End Try
            Else
                result = Nothing
                specified = False
            End If

            Return result
        End Function

#End Region

    End Module

End Namespace

File: Common/InMotionGIT.Common.Core/Extensions/DataRowExtensions.cs
using System;
using System.Data;

#region using

using System.IO;
using System.Xml.Linq;
using Microsoft.VisualBasic.CompilerServices;

#endregion using

namespace InMotionGIT.Common.Core.Extensions;

/// <summary>
/// Extension methods for the DataRow type
/// </summary>
public static class DataRowExtensions
{
    #region Binary (BLOB) Extension

    public static string FileContent(this DataRow value, string name, string path, string filename, string extension)
    {
        string result = string.Empty;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            string localfilename = path;
            if (filename.IsEmpty())
            {
                localfilename += @"\" + Guid.NewGuid().ToString();
            }
            else
            {
                localfilename += @"\" + filename;
            }
            if (extension.StartsWith("."))
            {
                localfilename += extension;
            }
            else
            {
                localfilename += "." + extension;
            }
            localfilename = localfilename.Replace(@"\\", @"\");

            using (var FS = new FileStream(localfilename, FileMode.Create))
            {
                byte[] blob = (byte[])value[name];
                FS.Write(blob, 0, blob.Length);
                FS.Close();
            }

            result = localfilename;
        }

        return result;
    }

    #endregion Binary (BLOB) Extension

    #region Numeric Extension

    public static decimal NumericValue(this DataRow value, string name)
    {
        decimal result = 0m;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToDecimal(value[name]);
        }

        return result;
    }

    public static int IntegerValue(this DataRow value, string name)
    {
        int result = 0;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToInteger(value[name]);
        }

        return result;
    }

    public static byte ByteValue(this DataRow value, string name)
    {
        byte result = 0;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToByte(value[name]);
        }

        return result;
    }

    public static object NumericValueWithFormatDefault(this DataRow value, string name, string format, object defaultValue, ref bool specified)
    {
        var result = defaultValue;

        if (value.IsNotNull(name))
        {
            decimal @internal = value.NumericValue(name);

            if (format.IsNotEmpty())
            {
                result = @internal.ToString(format, new System.Globalization.CultureInfo("en-US", false));
            }
            else
            {
                result = @internal;
            }

            specified = true;
        }
        else
        {
            specified = false;
        }

        return result;
    }

    #endregion Numeric Extension

    #region Double Extension

    public static double DoubleValue(this DataRow value, string name)
    {
        double result = 0d;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToDouble(value[name]);
        }

        return result;
    }

    #endregion Double Extension

    #region DateTime Extension

    public static DateTime DateTimeValue(this DataRow value, string name)
    {
        var result = DateTime.MinValue;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToDate(value[name]);
        }

        return result;
    }

    public static object DateTimeValueWithFormatDefault(this DataRow value, string name, string format, object defaultValue)
    {
        var result = defaultValue;

        if (value.IsNotNull(name))
        {
            var @internal = value.DateTimeValue(name);

            if (format.IsNotEmpty())
            {
                result = @internal.ToString(format);
            }
            else
            {
                result = @internal;
            }
        }

        return result;
    }

    #endregion DateTime Extension

    #region Boolean Extension

    public static bool BooleanValue(this DataRow value, string name)
    {
        bool result = Conversions.ToBoolean(0);

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(value[name], 1, false));
        }

        return result;
    }

    public static bool BooleanCharValue(this DataRow value, string name)
    {
        bool result = Conversions.ToBoolean(0);

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(value[name], "1", false));
        }

        return result;
    }

    #endregion Boolean Extension

    #region String Extension

    public static string StringValue(this DataRow value, string name)
    {
        string result = string.Empty;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToString(value[name]);
            result = result.Trim();
        }

        return result;
    }

    public static string StringValueWithDefault(this DataRow value, string name, object defaultValue, ref bool specified)
    {
        string result = Conversions.ToString(defaultValue);

        if (value.IsNotNull(name))
        {
            result = value.StringValue(name);
            specified = true;
        }
        else
        {
            specified = false;
        }

        return result;
    }

    public static DateTime StringHourValue(this DataRow value, string name)
    {
        var result = DateTime.MinValue;
        int currentHour = 0;
        int currentMinute = 0;
        string currentValue;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            currentValue = Conversions.ToString(value[name]);
            currentValue = currentValue.Trim();
            if (currentValue.IndexOf(":") > 0)
            {
                currentHour = Conversions.ToInteger(currentValue.Split(':')[0]);
                currentMinute = Conversions.ToInteger(currentValue.Split(':')[1]);
            }
        }

        result = result.AddHours(currentHour);
        result = result.AddMinutes(currentMinute);

        return result;
    }

    #endregion String Extension

    #region Specials Extension

    public static Guid GuidValue(this DataRow value, string name)
    {
        var result = Guid.NewGuid();

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = (Guid)value[name];
        }

        return result;
    }

    public static XDocument XmlValue(this DataRow value, string name)
    {
        var result = new XDocument();

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = XDocument.Parse(value.StringValue(name));
        }

        return result;
    }

    public static bool SwitchValue(this DataRow value, string name)
    {
        bool result = false;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(value[name], 1, false));
        }

        return result;
    }

    public static bool SwitchCharValue(this DataRow value, string name)
    {
        bool result = false;

        if (!(value[name] == null) && !(value[name] is DBNull))
        {
            result = Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(value[name], "1", false));
        }

        return result;
    }

    #endregion Specials Extension

    #region Behavior Functions

    public static bool IsNotNull(this DataRow value, string name)
    {
        return !(value[name] is DBNull);
    }

    public static T EnumValue<T>(this DataRow value, string name, T enumType, ref bool specified, ref bool witherror)
    {
        T result = default;

        if (value.IsNotNull(name))
        {
            try
            {
                result = Conversions.ToGenericParameter<T>(Enum.Parse(enumType.GetType(), Conversions.ToString(value[name])));
                specified = true;
            }
            catch (ArgumentException ex)
            {
                result = default;
                specified = false;
                witherror = true;
            }
        }
        else
        {
            result = default;
            specified = false;
        }

        return result;
    }

    #endregion Behavior Functions
}